contracts/storage/ServiceAgreementStorage.sol

```diff
@@ -0,0 +1,503 @@
+            timeNow < (agreement.startTime + agreement.epochLength * epoch + proofWindowOffset + proofWindowDuration)
+        );
+    }
+
+    function getChallenge(address assetContract, uint256 tokenId, bytes memory keyword, uint8 hashingAlgorithm)
+        public
+        returns (bytes32, uint256)
+    {
+        bytes32 agreementId = _generateAgreementId(assetContract, tokenId, keyword, hashingAlgorithm);
+
+        IdentityStorage identityStorage = IdentityStorage(hub.getContractAddress("IdentityStorage"));
+        uint96 identityId = identityStorage.identityIds(keccak256(abi.encodePacked(msg.sender)));
+
+        AbstractAsset generalAssetInterface = AbstractAsset(assetContract);
+        bytes32 assertionId = generalAssetInterface.getAssertionByIndex(
+            tokenId,
+            AbstractAsset(assetContract).getAssertionsLength(tokenId) - 1
+        );
+
+        AssertionRegistry assertionRegistry = AssertionRegistry(hub.getContractAddress("AssertionRegistry"));
+        uint256 assertionSize = assertionRegistry.getSize(assertionId);
+
+        // blockchash() function only works for last 256 blocks (25.6 min window in case of 6s block time)
+        // TODO: figure out how to achieve randomness
+        return (
+            assertionId,
+            uint256(
+                sha256(abi.encodePacked(serviceAgreements[agreementId].proofWindowOffsetPerc, identityId))
+            ) % assertionSize
+        );
+    }
+
+    function sendProof(
+        address assetContract,
+        uint256 tokenId,
+        bytes memory keyword,
+        uint8 hashingFunctionId,
+        uint16 epoch,
+        bytes32[] memory proof,
+        bytes32 chunkHash
+    )
+        public
+    {
+        bytes32 agreementId = _generateAgreementId(assetContract, tokenId, keyword, hashingFunctionId);
+        require(!isProofWindowOpen(agreementId, epoch), "Proof window is open");
+
+        uint96 identityId = IdentityStorage(hub.getContractAddress("IdentityStorage")).identityIds(
+            keccak256(abi.encodePacked(msg.sender))
+        );
+
+        require(
+            commitSubmissions[keccak256(abi.encodePacked(agreementId, epoch, identityId))].score != 0,
+            "You have been already rewarded in this epoch"
+        );
+
+        bytes32 nextCommitId = serviceAgreements[agreementId].epochSubmissionHeads[epoch];
+
+        ParametersStorage parametersStorage = ParametersStorage(hub.getContractAddress("ParametersStorage"));
+
+        bool isRewarded = false;
+        for (uint32 i = 0; i < parametersStorage.R0(); i++) {
+            CommitSubmission memory commit = commitSubmissions[nextCommitId];
+
+            if (identityId == commit.identityId) isRewarded = true;
+
+            nextCommitId = keccak256(abi.encodePacked(agreementId, epoch, commit.nextIdentity));
+        }
+
+        require(!isRewarded, "You hasn't been chosen for reward in this epoch!");
```

Comment:
probably better like this
"You're node hasn't been awarded for this asset in this epoch"